aea/aea_builder.py

```diff
@@ -0,0 +1,736 @@
+                "Agent configuration file '{}' is invalid. Please check the documentation.".format(
+                    DEFAULT_AEA_CONFIG_FILE
+                )
+            )
+
+    @classmethod
+    def from_aea_project(cls, aea_project_path: PathLike):
+        """
+        Construct the builder from an AEA project
+
+        - load agent configuration file
+        - set name and default configurations
+        - load private keys
+        - load ledger API configurations
+        - set default ledger
+        - load every component
+
+        :param aea_project_path: path to the AEA project.
+        :return: an AEA agent.
+        """
+        aea_project_path = Path(aea_project_path)
+        cls._try_to_load_agent_configuration_file(aea_project_path)
+        _verify_or_create_private_keys(aea_project_path)
+        builder = AEABuilder(with_default_packages=False)
+
+        # TODO isolate environment
+        # load_env_file(str(aea_config_path / ".env"))
+
+        # load agent configuration file
+        configuration_file = aea_project_path / DEFAULT_AEA_CONFIG_FILE
+
+        loader = ConfigLoader.from_configuration_type(ConfigurationType.AGENT)
+        agent_configuration = loader.load(configuration_file.open())
+
+        # set name and other configurations
+        builder.set_name(agent_configuration.name)
+        builder.set_default_ledger_api_config(agent_configuration.default_ledger)
+        builder.set_default_connection(
+            PublicId.from_str(agent_configuration.default_connection)
+        )
+
+        # load private keys
+        for (
+            ledger_identifier,
+            private_key_path,
+        ) in agent_configuration.private_key_paths_dict.items():
+            builder.add_private_key(ledger_identifier, private_key_path)
+
+        # load ledger API configurations
+        for (
+            ledger_identifier,
+            ledger_api_conf,
+        ) in agent_configuration.ledger_apis_dict.items():
+            builder.add_ledger_api_config(ledger_identifier, ledger_api_conf)
+
+        component_ids = itertools.chain(
+            [
+                ComponentId(ComponentType.PROTOCOL, p_id)
+                for p_id in agent_configuration.protocols
+            ],
+            [
+                ComponentId(ComponentType.CONNECTION, p_id)
+                for p_id in agent_configuration.connections
+            ],
+            [
+                ComponentId(ComponentType.SKILL, p_id)
+                for p_id in agent_configuration.skills
+            ],
+        )
+        for component_id in component_ids:
+            component_path = cls._find_component_directory_from_component_id(
+                aea_project_path, component_id
+            )
+            builder.add_component(component_id.component_type, component_path)
+
+        return builder
+
+
+def _verify_or_create_private_keys(aea_project_path: Path) -> None:
+    """Verify or create private keys."""
```

Comment:
This is copied from `aea.cli.common._verify_or_create_private_keys`, with some variation. It is used in `AEABuilder.from_aea_config` in order to emulate the behaviour of `aea run`.  In particular, the CLI command instantiates the private key if it is not in the project path. Of course it can be further improved and refactored.